File: mobile/android/app/src/main/java/com/agentcall/app/call/WebRTCClient.kt
package com.agentcall.app.call

import android.content.Context
import org.webrtc.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class WebRTCClient @Inject constructor(
    private val context: Context
) {
    private var peerConnectionFactory: PeerConnectionFactory? = null
    private var peerConnection: PeerConnection? = null
    private var audioTrack: AudioTrack? = null
    private var audioSource: AudioSource? = null
    private var eglBase: EglBase? = null

    private var onIceCandidate: ((IceCandidate) -> Unit)? = null
    private var onIceConnectionChange: ((PeerConnection.IceConnectionState) -> Unit)? = null
    private var onConnectionStateChange: ((PeerConnection.PeerConnectionState) -> Unit)? = null

    fun initialize() {
        PeerConnectionFactory.initialize(
            PeerConnectionFactory.InitializationOptions.builder(context)
                .setFieldTrials("")
                .createInitializationOptions()
        )

        eglBase = EglBase.create()

        val options = PeerConnectionFactory.Options()
        options.networkIgnoreMask = 0

        peerConnectionFactory = PeerConnectionFactory.builder()
            .setOptions(options)
            .createPeerConnectionFactory()
    }

    fun createPeerConnection(
        iceServers: List<PeerConnection.IceServer>,
        onIceCandidate: (IceCandidate) -> Unit,
        onIceConnectionChange: (PeerConnection.IceConnectionState) -> Unit,
        onConnectionStateChange: (PeerConnection.PeerConnectionState) -> Unit
    ): PeerConnection? {
        this.onIceCandidate = onIceCandidate
        this.onIceConnectionChange = onIceConnectionChange
        this.onConnectionStateChange = onConnectionStateChange

        val config = PeerConnection.RTCConfiguration(iceServers).apply {
            sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN
            continualGatheringPolicy = PeerConnection.ContinualGatheringPolicy.GATHER_CONTINUALLY
            iceCandidatePoolSize = 10
            bundlePolicy = PeerConnection.BundlePolicy.MAXBUNDLE
            rtcpMuxPolicy = PeerConnection.RtcpMuxPolicy.REQUIRE
            iceConnectionState = PeerConnection.IceConnectionState.CHECKING
            keyType = PeerConnection.KeyType.ECDSA
        }

        peerConnection = peerConnectionFactory?.createPeerConnection(config, object : PeerConnection.Observer {
            override fun onIceCandidate(candidate: IceCandidate) {
                onIceCandidate(candidate)
            }

            override fun onIceCandidatesRemoved(candidates: Array<IceCandidate>) {}

            override fun onSignalingChange(state: PeerConnection.SignalingState) {}

            override fun onIceConnectionChange(state: PeerConnection.IceConnectionState) {
                onIceConnectionChange(state)
            }

            override fun onIceConnectionReceivingChange(receiving: Boolean) {}

            override fun onIceGatheringChange(state: PeerConnection.IceGatheringState) {}

            override fun onAddStream(stream: MediaStream) {}

            override fun onAddTrack(track: RtpReceiver, streams: Array<MediaStream>) {}

            override fun onRemoveStream(stream: MediaStream) {}

            override fun onDataChannel(channel: DataChannel) {}

            override fun onRenegotiationNeeded() {}

            override fun onStandardizedIceConnectionChange(state: PeerConnection.IceConnectionState) {
                onIceConnectionChange(state)
            }

            override fun onConnectionChange(state: PeerConnection.PeerConnectionState) {
                onConnectionStateChange(state)
            }
        })

        return peerConnection
    }

    fun startAudio(): AudioTrack? {
        val constraints = MediaConstraints().apply {
            mandatory.add(MediaConstraints.KeyValuePair("echoCancellation", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("noiseSuppression", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("autoGainControl", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("googEchoCancellation", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("googAutoGainControl", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("googNoiseSuppression", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("googHighpassFilter", "true"))
        }

        audioSource = peerConnectionFactory?.createAudioSource(constraints)
        audioTrack = peerConnectionFactory?.createAudioTrack("audio0", audioSource)
        return audioTrack
    }

    fun createOffer(sdpObserver: SdpObserver) {
        val constraints = MediaConstraints().apply {
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "false"))
        }
        peerConnection?.createOffer(sdpObserver, constraints)
    }

    fun createAnswer(sdpObserver: SdpObserver) {
        val constraints = MediaConstraints().apply {
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "false"))
        }
        peerConnection?.createAnswer(sdpObserver, constraints)
    }

    fun setRemoteDescription(sdpObserver: SdpObserver, sdp: SessionDescription) {
        peerConnection?.setRemoteDescription(sdpObserver, sdp)
    }

    fun addIceCandidate(candidate: IceCandidate) {
        peerConnection?.addIceCandidate(candidate)
    }

    fun mute(muted: Boolean) {
        audioTrack?.setEnabled(!muted)
    }

    fun dispose() {
        peerConnection?.dispose()
        audioTrack?.dispose()
        audioSource?.dispose()
        peerConnectionFactory?.dispose()
        eglBase?.release()
        peerConnection = null
        audioTrack = null
        audioSource = null
        peerConnectionFactory = null
        eglBase = null
    }
}

File: mobile/ios-archived/AgentCall/AgentCall/Call/WebRTCClient.swift
import Foundation
import WebRTC

actor WebRTCClient {
    static let shared = WebRTCClient()
    private var factory: RTCPeerConnectionFactory?
    private var peerConnection: RTCPeerConnection?
    private var audioTrack: RTCAudioTrack?
    private var audioSource: RTCAudioSource?

    // Delegate callbacks
    private var onIceCandidate: ((RTCIceCandidate) -> Void)?
    private var onIceConnectionChange: ((RTCIceConnectionState) -> Void)?
    private var onConnectionStateChange: ((RTCPeerConnectionState) -> Void)?

    private init() {}

    func initialize() {
        RTCInitializeSSL()
        let encoderFactory = RTCDefaultVideoEncoderFactory()
        let decoderFactory = RTCDefaultVideoDecoderFactory()
        factory = RTCPeerConnectionFactory(encoderFactory: encoderFactory, decoderFactory: decoderFactory)
    }

    func createPeerConnection(
        turnUsername: String,
        turnCredential: String,
        stunHost: String = "turn.agentcall.example.com:3478",
        turnHost: String = "turn.agentcall.example.com:5349",
        onIceCandidate: @escaping (RTCIceCandidate) -> Void = { _ in },
        onIceConnectionChange: @escaping (RTCIceConnectionState) -> Void = { _ in },
        onConnectionStateChange: @escaping (RTCPeerConnectionState) -> Void = { _ in }
    ) -> RTCPeerConnection? {
        self.onIceCandidate = onIceCandidate
        self.onIceConnectionChange = onIceConnectionChange
        self.onConnectionStateChange = onConnectionStateChange

        let config = RTCConfiguration()
        config.sdpSemantics = .unifiedPlan
        config.iceCandidatePoolSize = 10
        config.bundlePolicy = .maxBundle
        config.rtcpMuxPolicy = .require
        config.continualGatheringPolicy = .gatherContinually

        let stun = RTCIceServer(urlStrings: ["stun:\(stunHost)"])
        let turn = RTCIceServer(
            urlStrings: ["turn:\(turnHost)"],
            username: turnUsername,
            credential: turnCredential
        )
        config.iceServers = [stun, turn]

        let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
        peerConnection = factory?.peerConnection(
            with: config,
            constraints: constraints,
            delegate: PeerConnectionDelegateProxy(
                onIceCandidate: { [weak self] candidate in
                    Task { await self?.onIceCandidate?(candidate) }
                },
                onIceConnectionChange: { [weak self] state in
                    Task { await self?.onIceConnectionChange?(state) }
                },
                onConnectionStateChange: { [weak self] state in
                    Task { await self?.onConnectionStateChange?(state) }
                }
            )
        )
        return peerConnection
    }

    func startAudio() -> RTCAudioTrack? {
        let constraints = RTCMediaConstraints(
            mandatoryConstraints: [
                "googEchoCancellation": "true",
                "googAutoGainControl": "true",
                "googNoiseSuppression": "true",
            ],
            optionalConstraints: nil
        )
        audioSource = factory?.audioSource(with: constraints)
        audioTrack = factory?.audioTrack(with: audioSource!, trackId: "audio0")
        peerConnection?.add(audioTrack!, streamIds: ["stream0"])
        return audioTrack
    }

    func createOffer() async throws -> RTCSessionDescription {
        return try await withCheckedThrowingContinuation { continuation in
            let constraints = RTCMediaConstraints(
                mandatoryConstraints: ["OfferToReceiveAudio": "true", "OfferToReceiveVideo": "false"],
                optionalConstraints: nil
            )
            peerConnection?.offer(for: constraints) { sdp, error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else if let sdp = sdp {
                    self.peerConnection?.setLocalDescription(sdp) { error in
                        if let error = error {
                            continuation.resume(throwing: error)
                        } else {
                            continuation.resume(returning: sdp)
                        }
                    }
                }
            }
        }
    }

    func createAnswer() async throws -> RTCSessionDescription {
        return try await withCheckedThrowingContinuation { continuation in
            let constraints = RTCMediaConstraints(
                mandatoryConstraints: ["OfferToReceiveAudio": "true", "OfferToReceiveVideo": "false"],
                optionalConstraints: nil
            )
            peerConnection?.answer(for: constraints) { sdp, error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else if let sdp = sdp {
                    self.peerConnection?.setLocalDescription(sdp) { error in
                        if let error = error {
                            continuation.resume(throwing: error)
                        } else {
                            continuation.resume(returning: sdp)
                        }
                    }
                }
            }
        }
    }

    func setRemoteDescription(_ sdp: RTCSessionDescription) async throws {
        return try await withCheckedThrowingContinuation { continuation in
            peerConnection?.setRemoteDescription(sdp) { error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else {
                    continuation.resume()
                }
            }
        }
    }

    func addIceCandidate(_ candidate: RTCIceCandidate) {
        peerConnection?.add(candidate)
    }

    func mute(_ muted: Bool) {
        audioTrack?.isEnabled = !muted
    }

    func dispose() {
        peerConnection?.close()
        peerConnection = nil
        audioTrack = nil
        audioSource = nil
    }
}

// MARK: - RTCPeerConnectionDelegate Proxy

private class PeerConnectionDelegateProxy: NSObject, RTCPeerConnectionDelegate {
    private let onIceCandidate: (RTCIceCandidate) -> Void
    private let onIceConnectionChange: (RTCIceConnectionState) -> Void
    private let onConnectionStateChange: (RTCPeerConnectionState) -> Void

    init(
        onIceCandidate: @escaping (RTCIceCandidate) -> Void,
        onIceConnectionChange: @escaping (RTCIceConnectionState) -> Void,
        onConnectionStateChange: @escaping (RTCPeerConnectionState) -> Void
    ) {
        self.onIceCandidate = onIceCandidate
        self.onIceConnectionChange = onIceConnectionChange
        self.onConnectionStateChange = onConnectionStateChange
        super.init()
    }

    func peerConnection(_ peerConnection: RTCPeerConnection, didChange state: RTCSignalingState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {}

    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState) {
        onIceConnectionChange(newState)
    }

    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCPeerConnectionState) {
        onConnectionStateChange(newState)
    }

    func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        onIceCandidate(candidate)
    }

    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {}
    func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) -> Bool { return false }

    func peerConnection(_ peerConnection: RTCPeerConnection, didStartReceivingOn transceiver: RTCRtpTransceiver) {}
}
